Change SHOW COLUMNS query to display MySQL types in MySQL Compatibility mode

tpanetti opened the pull request:
This updates the SHOW COLUMN SQL query to display MySQL types when this query is issued by a client connected via MySQL Compatibility port

<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- New Feature


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add a new setting named `use_mysql_types_in_show_columns` to alter the `SHOW COLUMNS` SQL statement to display MySQL equivalent types when a client is connected via the MySQL compatibility port. 

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?
- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.
No arguments or parameters are needed, this type conversion is done when a client is connected via the MySQL port. Specifically, the implementation happens when the client is connected to whatever port the configuration has set for MySQL compatibility.
- Example use: A query or command.
-->
Motivation: This functionality is useful for ClickHouse users as it allows tools that use standard MySQL connectors to read and parse ClickHouse data as if it were MySQL. This is particularly useful for data analytics tools such as QuickSights and Looker Studio.
Parameters: No arguments or parameters are needed, this type conversion is done when a client is connected via the MySQL port. Specifically, the implementation happens when the client is connected to whatever port the configuration has set for MySQL compatibility.
Example use:
`SHOW FULL COLUMNS`

I've attached an image of Amazon QuickSight parsing a ClickHouse database with example cell_towers data with this code change.
> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/
<img width="1728" alt="Screenshot 2023-05-05 at 12 13 27 PM" src="https://user-images.githubusercontent.com/10793401/236560767-b1c3afda-3b93-4318-86ac-71963b07f7f6.png">

Closes [#49168](https://github.com/ClickHouse/ClickHouse/issues/49168)

tpanetti committed:
Change SHOW COLUMNS query to display MySQL types in MySQL Compatibility mode

This updates the SHOW COLUMN SQL query to display MySQL types when this
query is issued by a client connected via MySQL Compatibility port

tpanetti committed:
Change SHOW COLUMNS query to display MySQL types in MySQL Compatibility mode

This updates the SHOW COLUMN SQL query to display MySQL types when this
query is issued by a client connected via MySQL Compatibility port

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=49577) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=49577) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=49577) it.</sub>

alexey-milovidov commented on src/Interpreters/InterpreterShowColumnsQuery.cpp:
```diff
@@ -93,6 +99,74 @@ String InterpreterShowColumnsQuery::getRewrittenQuery()
+{
+    WriteBufferFromOwnString mysql_specific_query;
+
+    mysql_specific_query << "SELECT name AS field, "
```
It does not look maintainable.

alexey-milovidov commented on src/Interpreters/InterpreterShowColumnsQuery.cpp:
```diff
@@ -93,6 +99,74 @@ String InterpreterShowColumnsQuery::getRewrittenQuery()
+        << "CASE "
+        << "  WHEN startsWith(type, 'Nullable') THEN "
+        << "    CASE "
+        << "      WHEN substring(type, 10, length(type) - 10) IN ('UInt8', 'Int8') THEN 'tinyint' "
```
The code does not explain what 10 is.

alexey-milovidov commented on src/Interpreters/InterpreterShowColumnsQuery.cpp:
```diff
@@ -93,6 +99,74 @@ String InterpreterShowColumnsQuery::getRewrittenQuery()
+        << "  WHEN startsWith(type, 'Nullable') THEN "
+        << "    CASE "
+        << "      WHEN substring(type, 10, length(type) - 10) IN ('UInt8', 'Int8') THEN 'tinyint' "
+        << "      WHEN substring(type, 10, length(type) - 10) IN ('UInt16', 'Int16') THEN 'smallint' "
```
This introduces an implicit dependency between this file and every file inside the DataTypes directory.

tpanetti replied:
This is a good point. Perhaps we should instead build this string by iterating over the TypeIndex enum for all of the ClickHouse types and get the equivalent MySQL type with a function getSQLCompatibleName in IDataType (as you mentioned before), then look it up in a type equivalency map. That should solve the implicit dependency and make this more manageable although we would need to maintain the equivalency map.

alexey-milovidov commented on src/Interpreters/InterpreterShowColumnsQuery.cpp:
```diff
@@ -93,6 +99,74 @@ String InterpreterShowColumnsQuery::getRewrittenQuery()
+        << "      WHEN substring(type, 10, length(type) - 10) IN ('UInt8', 'Int8') THEN 'tinyint' "
+        << "      WHEN substring(type, 10, length(type) - 10) IN ('UInt16', 'Int16') THEN 'smallint' "
+        << "      WHEN substring(type, 10, length(type) - 10) IN ('UInt32', 'Int32') THEN 'int' "
+        << "      WHEN substring(type, 10, length(type) - 10) IN ('UInt64', 'Int64', 'UInt128', 'Int128', 'UInt256', 'Int256') THEN 'bigint' "
```
This is wrong because MySQL's bigint is just 64 bits.

alexey-milovidov commented on src/Interpreters/InterpreterShowColumnsQuery.cpp:
```diff
@@ -93,6 +99,74 @@ String InterpreterShowColumnsQuery::getRewrittenQuery()
+        << "      WHEN substring(type, 10, length(type) - 10) = 'Float64' THEN 'double' "
+        << "      WHEN substring(type, 10, length(type) - 10) LIKE 'Decimal%' THEN 'decimal' "
+        << "      WHEN substring(type, 10, length(type) - 10) = 'Boolean' THEN 'tinyint' "
+        << "      WHEN substring(type, 10, length(type) - 10) = 'String' THEN 'text' "
```
String can contain arbitrary binary data.

alexey-milovidov commented on src/Interpreters/InterpreterShowColumnsQuery.cpp:
```diff
@@ -93,6 +99,74 @@ String InterpreterShowColumnsQuery::getRewrittenQuery()
+        << "      WHEN substring(type, 10, length(type) - 10) LIKE 'FixedString%' THEN 'text' "
+        << "      WHEN substring(type, 10, length(type) - 10) LIKE 'Date%' THEN 'date' "
+        << "      WHEN substring(type, 10, length(type) - 10) LIKE 'DateTime%' THEN 'datetime' "
+        << "      WHEN substring(type, 10, length(type) - 10) = 'JSON' THEN 'json' "
```
This does not look correct.

alexey-milovidov commented on src/Interpreters/InterpreterShowColumnsQuery.cpp:
```diff
@@ -93,6 +99,74 @@ String InterpreterShowColumnsQuery::getRewrittenQuery()
+        << "      WHEN substring(type, 10, length(type) - 10) LIKE 'Date%' THEN 'date' "
+        << "      WHEN substring(type, 10, length(type) - 10) LIKE 'DateTime%' THEN 'datetime' "
+        << "      WHEN substring(type, 10, length(type) - 10) = 'JSON' THEN 'json' "
+        << "      WHEN substring(type, 10, length(type) - 10) = 'UUID' THEN 'binary' "
```
Why binary?

alexey-milovidov commented on src/Interpreters/InterpreterShowColumnsQuery.cpp:
```diff
@@ -93,6 +99,74 @@ String InterpreterShowColumnsQuery::getRewrittenQuery()
+        << "      WHEN substring(type, 10, length(type) - 10) = 'JSON' THEN 'json' "
+        << "      WHEN substring(type, 10, length(type) - 10) = 'UUID' THEN 'binary' "
+        << "      WHEN substring(type, 10, length(type) - 10) LIKE 'Enum%' THEN 'enum' "
+        << "      WHEN substring(type, 10, length(type) - 10) LIKE 'LowCardinality%' THEN 'text' "
```
This is wrong.

alexey-milovidov commented on src/Interpreters/InterpreterShowColumnsQuery.cpp:
```diff
@@ -93,6 +99,74 @@ String InterpreterShowColumnsQuery::getRewrittenQuery()
+        << "      WHEN substring(type, 10, length(type) - 10) = 'UUID' THEN 'binary' "
+        << "      WHEN substring(type, 10, length(type) - 10) LIKE 'Enum%' THEN 'enum' "
+        << "      WHEN substring(type, 10, length(type) - 10) LIKE 'LowCardinality%' THEN 'text' "
+        << "      WHEN substring(type, 10, length(type) - 10) LIKE 'Array%' THEN 'json' "
```
I'm not sure that this will work.

tpanetti committed:
Convert Clickhouse Types to MySQL types in Compatibility mode

This changes MySQL compatibility mode to display MySQL compatible types

tpanetti committed:
Merge branch 'master' into mysql_compatible_types

tpanetti committed:
Merge branch 'mysql_compatible_types' of https://github.com/tpanetti/ClickHouse into mysql_compatible_types

tpanetti committed:
Revert "Change SHOW COLUMNS query to display MySQL types in MySQL Compatibility mode"

This reverts commit ddbad79c5e67518acebbacaad5be0cad3967ac67.

tpanetti committed:
Fix incompatible ClickHouse -> MySQL types for compability mode

This adjusts specific incompatible ClickHouse types to a format that can
be read and interpreted by MySQL (Ex: Int128 -> text)

tpanetti commented:
Refactored to a more scalable method of implementation via the IDataType interface. Added @rschu1ze as an additional reviewer after discussions with him about this change. For incompatible types (eg int64, int128) I chose to default to text for compatibility with BI tools, this may warrant further discussion though.

alexey-milovidov commented:
Ok. Need to fix the test. Copy-paste the initialization header from similar .sh tests.

alexey-milovidov commented:
Let's rename `getMySQLName` to `getSQLCompatibleName`, and change the returned value to ALL CAPS (`TEXT`, etc).

tpanetti committed:
Merge branch 'master' into mysql_compatible_types

tpanetti committed:
Change case and function name for MySQL Compatible types

This changes the function name for MySQL Compatible types from getMySQLName to getSQLCompatibleName and changes the casing of the types to upper

tpanetti commented:
Added the initialization header from similar .sh tests, fixed the casing of the MySQL types and renamed `getMySQLName` to `getSQLCompatibleName`

alexey-milovidov commented on src/DataTypes/DataTypeEnum.h:
```diff
@@ -45,13 +45,16 @@ class DataTypeEnum final : public IDataTypeEnum, public EnumValues<Type>
 
     std::string doGetName() const override { return type_name; }
     const char * getFamilyName() const override;
+    const char * getSQLCompatibleName() const override { return my_sql_type_name.c_str(); }
```
Now returning const char * starts to be dangerous, let's change the interface of this function to return String.

Also, let's generate it on the fly - do not store it in the class's member.

alexey-milovidov commented on src/DataTypes/DataTypeLowCardinality.h:
```diff
@@ -22,6 +24,8 @@ class DataTypeLowCardinality : public IDataType
         return "LowCardinality(" + dictionary_type->getName() + ")";
     }
     const char * getFamilyName() const override { return "LowCardinality"; }
+    const char * getSQLCompatibleName() const override { return mysql_name.c_str(); }
```
.

alexey-milovidov commented on src/DataTypes/DataTypeNumberBase.h:
```diff
@@ -20,11 +20,14 @@ class DataTypeNumberBase : public IDataType
     using ColumnType = ColumnVector<T>;
 
     const char * getFamilyName() const override { return TypeName<T>.data(); }
+    const char * getSQLCompatibleName() const override { return mysqlTypeMap.at(TypeName<T>.data()).c_str(); }
```
We should avoid the usage of TypeName. I think writing a series of `if constexpr` directly here will be better.

alexey-milovidov commented on src/DataTypes/DataTypeString.h:
```diff
@@ -21,6 +21,9 @@ class DataTypeString final : public IDataType
         return "String";
     }
 
+    // FIXME: string can contain arbitrary bytes, not only UTF-8 sequences
```
Nothing to fix here.

But it raises a question, why we are using TEXT for AggregateFunction - it contains non-UTF8 sequences almost always.

alexey-milovidov commented on src/DataTypes/DataTypeTuple.h:
```diff
@@ -33,6 +33,7 @@ class DataTypeTuple final : public IDataType
     TypeIndex getTypeId() const override { return TypeIndex::Tuple; }
     std::string doGetName() const override;
     const char * getFamilyName() const override { return "Tuple"; }
+    const char * getSQLCompatibleName() const override { return "JSON"; }
```
Not sure how external tools will eat it when we provide values of this type.

alexey-milovidov commented on src/DataTypes/DataTypeUUID.h:
```diff
@@ -18,6 +18,8 @@ class DataTypeUUID : public IDataType
     static constexpr auto type_id = TypeIndex::UUID;
 
     const char * getFamilyName() const override { return "UUID"; }
+    const char * getSQLCompatibleName() const override { return "CHAR"; }
```
CHAR in MySQL is of fixed size. Specifying CHAR without a size can lead to something strange.

alexey-milovidov replied:
Although it might work...

alexey-milovidov commented on src/DataTypes/IDataType.h:
```diff
@@ -71,10 +71,19 @@ class IDataType : private boost::noncopyable, public std::enable_shared_from_thi
+    /// MySQL equivalent Name of data type (examples: UInt64, Array(String)).
+    String getMySQLTypeName() const
+    {
+        if (custom_name)
```
I don't think that this branch is relevant.

alexey-milovidov commented on src/Storages/System/StorageSystemColumns.cpp:
```diff
@@ -74,6 +74,7 @@ class ColumnsSource : public ISource
         : ISource(header_)
         , columns_mask(std::move(columns_mask_)), max_block_size(max_block_size_)
         , databases(std::move(databases_)), tables(std::move(tables_)), storages(std::move(storages_))
+        , clientInfo(context->getClientInfo())
```
clientInfo -> client_info,
but it is better not to copy the whole struct here but only check for a particular member and set a flag.

alexey-milovidov commented on src/Storages/System/StorageSystemColumns.cpp:
```diff
@@ -129,6 +130,17 @@ class ColumnsSource : public ISource
+            {
+                if (clientInfo.interface == DB::ClientInfo::Interface::MYSQL)
+                {
+                   return type.getMySQLTypeName();
```
Wrong indent - three spaces instead of four.
Wrong function name.

alexey-milovidov commented:
If we change the behavior of the MySQL interface by default, it is going to be a backward incompatible change - let's make a setting to control its behavior.

rschu1ze commented on src/DataTypes/DataTypeFixedString.h:
```diff
@@ -42,6 +42,7 @@ class DataTypeFixedString final : public IDataType
     TypeIndex getTypeId() const override { return type_id; }
 
     const char * getFamilyName() const override { return "FixedString"; }
+    const char * getSQLCompatibleName() const override { return "TEXT"; }
```
Isn't that `CHAR` in MySQL? https://dev.mysql.com/doc/refman/8.0/en/char.html

EDIT: Ah, guess TEXT is more fitting because it can hold arbitrary bytes and CHAR can't. Maybe add a comment like in DataTypeString.h.

rschu1ze commented on src/DataTypes/DataTypesDecimal.h:
```diff
@@ -37,8 +37,11 @@ class DataTypeDecimal final : public DataTypeDecimalBase<T>
+    static constexpr auto mysql_name = "DECIMAL";
 
     const char * getFamilyName() const override { return family_name; }
+    const char * getSQLCompatibleName() const override { return mysql_name; }
```
It will be cleaner to include `M` and `D` parameters: https://dev.mysql.com/doc/refman/8.0/en/fixed-point-types.html

rschu1ze commented on tests/queries/0_stateless/02775_show_columns_mysql_compatibility.sh:
```diff
@@ -0,0 +1,129 @@
+#!/usr/bin/env bash
```
Curious why this test can't be a "normal" .sql test?

tpanetti replied:
Since we're testing both the clickhouse-client and mysql client simultaneously we need to use a shell script to leverage both tools.

rschu1ze commented on tests/queries/0_stateless/02775_show_columns_mysql_compatibility.sh:
```diff
@@ -0,0 +1,129 @@
+    SET allow_experimental_object_type =1;
+    CREATE TABLE tab
+    (
+        uint64 UInt64,
```
(Nitpicking: not all integer types are tested).

alexey-milovidov commented:
@rschu1ze VARCHAR/TEXT/BLOB, etc are SQL standard types, and we are mostly inside this set.

tpanetti committed:
Refactor ClickHouse->MySQL Type conversion and add configuration setting to trigger type conversion

tpanetti commented:
Addressed all comments, added the configuration setting named `output_format_mysql_types` to enable MySQL type conversion so we can drop the `pr-backward-incompatible` tag.

rschu1ze commented on src/DataTypes/IDataType.h:
```diff
@@ -75,6 +75,8 @@ class IDataType : private boost::noncopyable, public std::enable_shared_from_thi
 
     /// Name of data type family (example: FixedString, Array).
     virtual const char * getFamilyName() const = 0;
+    /// Name of corresponding data type in MySQL (exampe: Bigint, Blob, etc)
```
(For consistency, it would be great to have `getFamilyName()` also return a String. But that's unrelated and fine to be done in a separate PR.)

alexey-milovidov replied:
It's ok for `getFamilyName` to return a reference to a statically allocated string (const char *), because the family name is constant for the class.

rschu1ze commented on src/DataTypes/DataTypeNumberBase.cpp:
```diff
@@ -11,6 +11,55 @@ Field DataTypeNumberBase<T>::getDefault() const
+template <typename T>
+String DataTypeNumberBase<T>::getSQLCompatibleName() const
+{
+    if constexpr (std::is_same_v<T, Int8>)
```
(I don't care too much but parentheses for single-line if/for/while statements are usually omitted in the codebase. In this case, parentheses blow the code artificially up.)

rschu1ze reviewed (commented):
Lgtm, thanks.

tpanetti committed:
Merge branch 'master' into mysql_compatible_types

tpanetti committed:
Disable fasttest for MySQL Compatibility Type Conversion and refactor style for DataTypeNumberBase

tpanetti committed:
Merge branch 'master' into mysql_compatible_types

alexey-milovidov commented on src/Core/Settings.h:
```diff
@@ -196,6 +196,7 @@ class IColumn;
     M(Bool, allow_experimental_inverted_index, false, "If it is set to true, allow to use experimental inverted index.", 0) \
     \
     M(UInt64, mysql_max_rows_to_insert, 65536, "The maximum number of rows in MySQL batch insertion of the MySQL storage engine", 0) \
+    M(Bool, output_format_mysql_types, false, "Use MySQL converted types when connected via MySQL compatibility", 0) \
```
All LGTM except this setting name and description.

It does not force output MySQL types everywhere, only for SHOW COLUMNS. For example, it does not affect Native, JSON, and CSVWithNamesAndTypes output formats.

Let's rename it to `use_mysql_types_in_show_columns`?

tpanetti committed:
Rename setting and description for MySQL compatible types
This renames the setting for MySQL compatible types from output_format_mysql_types to use_mysql_types_in_show_columns

rschu1ze reviewed (approved):
Lgtm (thanks) but the tests still fail in a weird way.

E.g. here: https://s3.amazonaws.com/clickhouse-test-reports/49577/cb8c20722b8976fe0bc402498667b02c2585cc02/stateless_tests_flaky_check__asan_.html

I did not check deeper but there can be a race on database name `database_123456789abcde` when the test runs in parallel. You could use tag "no-parallel" at the top of the test file to suppress parallel execution.

tpanetti committed:
Add "no-parallel" tag to MySQL Compatible Types test to fix test issue

tpanetti committed:
Merge branch 'master' into mysql_compatible_types

tpanetti committed:
Merge branch 'master' into mysql_compatible_types

tpanetti committed:
Merge branch 'master' into mysql_compatible_types

tpanetti committed:
Merge branch 'master' into mysql_compatible_types

tpanetti committed:
Merge branch 'master' into mysql_compatible_types

Avogar commented on tests/queries/0_stateless/02775_show_columns_mysql_compatibility.sh:
```diff
@@ -0,0 +1,153 @@
+
+echo "Drop tables if they exist"
+${CLICKHOUSE_LOCAL} --query "DROP TABLE IF EXISTS tab"
+${CLICKHOUSE_LOCAL} --query "DROP TABLE IF EXISTS database_123456789abcde"
```
```suggestion
${CLICKHOUSE_LOCAL} --query "DROP DATABASE IF EXISTS database_123456789abcde"
```

tpanetti replied:
Fixed

tpanetti committed:
Fix test for MySQL Compatible Types to use clickhouse_client rather than clickhouse_local

tpanetti committed:
Fix test for MySQL Compatible types (drop database properly)

tpanetti committed:
Merge branch 'master' into mysql_compatible_types

tpanetti committed:
Fix test for MySQL Compatible Types (suppress password warning)

tpanetti committed:
Merge branch 'master' into mysql_compatible_types

rschu1ze commented:
ClickHouse Integration Tests (asan) [6/6]:
- test_ssl_cert_authentication: #49400

ClickHouse Stateless Tests (debug) [2/5]
- 02703_max_local_write_bandwidth #51221
- 02561_null_as_default_more_formats #49177

antonio2368 commented on tests/queries/0_stateless/02775_show_columns_mysql_compatibility.sh:
```diff
@@ -0,0 +1,153 @@
+
+echo "Drop tables if they exist"
+${CLICKHOUSE_CLIENT} --query "DROP TABLE IF EXISTS tab"
+${CLICKHOUSE_CLIENT} --query "DROP DATABASE IF EXISTS database_123456789abcdef"
```
`no-parallel` can be removed and `${CLICKHOUSE_DATABASE}` can be used to get the unique generated database for this test.
